refactor(core/cardano): decouple address parameters validation

gabrielKerekes opened the pull request:
Closes https://github.com/vacuumlabs/trezor-firmware/issues/59

refi93 commented on core/src/apps/cardano/address.py:
```diff
@@ -34,6 +39,39 @@
+    staking_path: list[int],
+    staking_key_hash: bytes | None,
+) -> None:
+    if (staking_key_hash is None) == (not staking_path):
```
I'd rather break this down into two separate condition asserting that neither both nor none can be present

gabrielKerekes replied:
Perhaps we don't need to do a validation of both of them being present? If you pass in `staking_key_hash` that is the one that's going to be used. Otherwise we validate and then also use `staking_path`?

I remember this specific condition being recommended by someone from Trezor in their review of the Shelley update, but I think we can separate and see if they accept it - if we decide to keep it.

gabrielKerekes replied:
If we don't validate the case of them both being present, then the whole function would be more readable. And I also noticed, that we should probably check the length of `staking_key_hash`.

janmazak replied:
I agree it should be rewritten to be more readable. `staking_key_hash` naturally splits it into two cases to handle.

gabrielKerekes replied:
I've updated the `if`s. It now doesn't check whether both `staking_key_hash` and `staking_path` are set. If `staking_key_hash` is set, that is the one that's used.

I've also added a check for `staking_key_hash` length and a test for that.

refi93 commented on core/src/apps/cardano/address.py:
```diff
@@ -34,6 +39,39 @@
+    if (staking_key_hash is None) == (not staking_path):
+        raise INVALID_ADDRESS_PARAMETERS
+
+    if staking_key_hash is None and not SCHEMA_STAKING_ANY_ACCOUNT.match(staking_path):
```
I'd find it more natural to check `staking_path is not None and ...` - the mutual exclusive presence of those parameters is being checked by the top assertion after all

gabrielKerekes replied:
See [comment](https://github.com/vacuumlabs/trezor-firmware/pull/60#discussion_r617306449) above.

janmazak commented on core/src/apps/cardano/helpers/__init__.py:
```diff
@@ -1,6 +1,7 @@
 
 INVALID_ADDRESS = wire.ProcessError("Invalid address")
+INVALID_ADDRESS_PARAMETERS = wire.ProcessError("Invalid address parameters")
 NETWORK_MISMATCH = wire.ProcessError("Output address network mismatch!")
```
Why is `!` just here?

gabrielKerekes replied:
True. Removed.

janmazak commented on core/src/apps/cardano/certificates.py:
```diff
@@ -7,7 +7,7 @@
     validate_reward_address,
 )
-from .helpers import INVALID_CERTIFICATE, LOVELACE_MAX_SUPPLY
+from .helpers import INVALID_CERTIFICATE, LOVELACE_MAX_SUPPLY, PUBLIC_KEY_HASH_SIZE
```
I've not checked across all the codebase, but having a basic definition such as `PUBLIC_KEY_HASH_SIZE` in `helpers` seems fishy. I would expect something more like `cardano.h` for ledger which groups all those external constants. Also, the name used there is `ADDRESS_KEY_HASH_LENGTH` --- there are other public keys around, e.g. VRF key is hashed to 32 bytes, so we want something more specific.

gabrielKerekes replied:
Agree, renamed in https://github.com/vacuumlabs/trezor-firmware/pull/60/commits/b5c1ee5fe843ec8daed762b157e6e8d7db49b168. I have kept it in the helpers module though. It doesn't seem to be a convetion to have constants in the `__init__.py` of the app (cardano) module (which is IMHO equivalent to `cardano.h`). If there were more constants used throughout the Cardano module we could introduce a `consts.py` file or something similar, but I don't think that's necessary for the two values we currently have.

gabrielKerekes commented:
In the end I decided to add the check that `staking_key_hash` and `staking_path` aren't present at the same time. https://github.com/vacuumlabs/trezor-firmware/pull/60/commits/8658d3d123eccdec4b9b59244da300049feb8269

I also added a commit for checking that outputs don't contain both `address` and `address_parameters`. We discussed this offline with @refi93 and came to the conlcusion that this check should be there. In order for this to be testable I also updated trezorlib to add both attributes if they are present. https://github.com/vacuumlabs/trezor-firmware/pull/60/commits/b0f1ca4787981c29c316b2b254a440b5741c88c7

gabrielKerekes commented:
@janmazak please take another look at this.

gabrielKerekes commented:
I've also added the commit which reintroduces max tx output size to this PR https://github.com/vacuumlabs/trezor-firmware/pull/60/commits/ac665182f504e88a3a53ced92e54ac3ce7e57f8e. I think it should be better for Trezor guys to have a single PR which contains multiple small fixes than three tiny PRs.

janmazak commented on core/src/apps/cardano/address.py:
```diff
@@ -193,9 +238,6 @@ def derive_address_bytes(
 def _derive_byron_address(
```
Is this function actually of any use? Let's call `derive_byron_address` directly.

gabrielKerekes replied:
https://github.com/vacuumlabs/trezor-firmware/pull/60/commits/f9fa839bf899aa479577b3a88bf9bb34621215ee

janmazak commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -213,11 +211,14 @@ def _validate_outputs(
-            derive_address_bytes(
-                keychain, output.address_parameters, protocol_magic, network_id
+
+        if output.address_parameters and output.address is not None:
```
Perhaps compare `output.address_parameters` with `None` explicitly?

gabrielKerekes replied:
I don't think this is needed as this is the way objects are checked throughout the project and also in the next condition. Address is checked against `None` since it's a string. I would keep it as is.

refi93 commented on core/src/apps/cardano/address.py:
```diff
@@ -193,9 +238,6 @@ def derive_address_bytes(
 def _derive_byron_address(
```
does this function make sense anymore? Without the validation it seems to act just as an alias for `derive_byron_address`

gabrielKerekes replied:
https://github.com/vacuumlabs/trezor-firmware/pull/60/commits/f9fa839bf899aa479577b3a88bf9bb34621215ee

janmazak reviewed (approved):
Only two minor things.

refi93 commented on core/src/apps/cardano/address.py:
```diff
@@ -34,6 +40,45 @@
 
 
+def validate_address_parameters(parameters: CardanoAddressParametersType) -> None:
+    if parameters.address_type == CardanoAddressType.BYRON:
```
to rule out any ambiguity, I'd enforce that the rest of fields of parameters should not be set, something like
```python
if any([parameters.address_n_staking, parameters.staking_key_hash]):
    raise INVALID_ADDRESS_PARAMETERS
```

to make it less cumbersome, the individual parameters may be extracted into separate variables so you don't have to prefix it with "params." all the time

gabrielKerekes replied:
What do you think of this - https://github.com/vacuumlabs/trezor-firmware/pull/60/commits/b09e312a9aa6e275b8ef1430822364757ce15cd4?

refi93 replied:
sorry for the delayed reply - looks good to me :)

refi93 commented on core/src/apps/cardano/auxiliary_data.py:
```diff
@@ -52,10 +51,7 @@ def validate_auxiliary_data(
-            auxiliary_data.catalyst_registration_parameters,
-            protocol_magic,
-            network_id,
+            auxiliary_data.catalyst_registration_parameters
```
I think the original logic ensured that the protocol magic/network id is consistent with the overall transaction network id/protocol magic - is this check still preserved?

refi93 replied:
I see - address parameters don't contain the network id/protocol magic, the top-level protocol magic/network id is automatically used when deriving from address params, so we don't really need to ensure this consistency here

refi93 commented on core/src/apps/cardano/helpers/__init__.py:
```diff
@@ -1,7 +1,8 @@
 INVALID_ADDRESS = wire.ProcessError("Invalid address")
-NETWORK_MISMATCH = wire.ProcessError("Output address network mismatch!")
+INVALID_ADDRESS_PARAMETERS = wire.ProcessError("Invalid address parameters")
+NETWORK_MISMATCH = wire.ProcessError("Output address network mismatch")
```
I think this error is relevant for Catalyst as well in which case I'd reword the message to "Address network mismatch"

refi93 replied:
I realized address params don't contain the network id/protocol magic so this is apparently relevant only for output addresses given in their raw form

refi93 reviewed (changes requested):
just a few improvement ideas and clarifying questions

janmazak commented on core/src/apps/cardano/address.py:
```diff
@@ -34,6 +40,71 @@
+        raise INVALID_ADDRESS_PARAMETERS
+
+
+def _validate_address_parameters_fields(
```
I find the name somewhat misleading, like if it was about individual fields. What about `_check_superfluous_fields` or something like that?

gabrielKerekes replied:
Maybe `_validate_address_parameters_structure`? `_validate_address_parameters_format`?

gabrielKerekes replied:
I've changeed it to `validate_address_parameters_structure` https://github.com/vacuumlabs/trezor-firmware/pull/60/commits/8b08dc8000fc77243a02062c76dfb8752eeee064.

refi93 commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -226,11 +227,39 @@ def _validate_outputs(
+    network_id: int,
+) -> None:
+    """
+    Output size is currently limited to 4000 bytes at protocol level. Given
```
I double checked the ledger specs and the memory limit apparently applies only to the "value" part of the utxo (i.e. not the address)

sources:

Ledger specs (notice "value" field is being extracted):
https://github.com/input-output-hk/cardano-ledger-specs/pull/2099/files#diff-1dbf0b3dcad4b1ddcdf86e9df16469b[…]461f3e9f906cf42d925d6fb3a386dd6c2R258

CDDL
https://github.com/input-output-hk/cardano-ledger-specs/blob/5e5b9a20f1b4818681996[…]e3859de49/shelley-ma/shelley-ma-test/cddl-files/shelley-ma.cddl

I also asked here for the value of the parameter for Alonzo era: https://github.com/input-output-hk/cardano-node/pull/2634#issuecomment-831111987

gabrielKerekes replied:
Done https://github.com/vacuumlabs/trezor-firmware/pull/60/commits/a09056ccb5056b0ae4ed10ee7130ec53441c1d73. The actual limit can be adjusted later when the parameter value in Alonzo will be known.

refi93 commented on core/src/apps/cardano/sign_tx.py:
```diff
@@ -226,11 +227,43 @@ def _validate_outputs(
+
+    if serialized_output_size > MAX_TX_OUTPUT_SIZE:
+        raise wire.ProcessError(
+            "Maximum tx output size (%s bytes) exceeded!" % MAX_TX_OUTPUT_SIZE
```
I'd update

```suggestion
            "Maximum tx output size (%s bytes) exceeded!" % MAX_TX_OUTPUT_SIZE
```
to

```suggestion
            "Maximum tx output value size (%s bytes) exceeded!" % MAX_TX_OUTPUT_SIZE
```

to reflect reality

gabrielKerekes commented:
Merged via https://github.com/trezor/trezor-firmware/pull/1606.
